KeenClient/KIODBStore.m

```diff
@@ -143,35 +157,45 @@ - (BOOL)deleteAndRecreateCorruptDB {
 }
 
 - (BOOL)openAndInitDB {
-    if(!dbIsOpen) {
-        if (![self openDB]) {
-            return false;
-        } else {
-            if(![self createTables]) {
-                KCLogError(@"Failed to create SQLite table!");
-                [self closeDB];
-                return false;
-            }
-        
-            if (![self migrateTable]) {
-                KCLogError(@"Failed to migrate SQLite table!");
-                [self closeDB];
-                return false;
-            }
-        
-            if (![self prepareAllSQLiteStatements]) {
+
+    // The database can be opened from the thread calling into the SDK,
+    // or by the database queue thread if it has been closed, so open
+    // needs to be protected with a lock.
+    [openLock lock];
```

Comment:
Have we gotten to a point where we expect this to be happening infrequently? If not, should we consider at some point a less resource-intensive? Or is NSLock already fairly nice about trying in user mode first before doing anything too drastic?